mlflow/types/schema.py

```diff
@@ -116,84 +118,189 @@ def __repr__(self) -> str:
+        """Return true iff this schema declares names, false otherwise. """
+        return self.inputs and self.inputs[0].name is not None
 
     def column_types(self) -> List[DataType]:
-        """ Get column types of the columns in the dataset."""
-        return [x.type for x in self._cols]
+        """ Get types of the represented dataset. Unsupported by TensorSpec."""
+        if self.is_tensor_spec():
+            raise MlflowException("TensorSpec only supports numpy types, use numpy_types() instead")
+        return [x.type for x in self.inputs]
 
     def numpy_types(self) -> List[np.dtype]:
         """ Convenience shortcut to get the datatypes as numpy types."""
-        return [x.type.to_numpy() for x in self.columns]
+        if self.is_tensor_spec():
+            return [x.type for x in self.inputs]
+        return [x.type.to_numpy() for x in self.inputs]
 
     def pandas_types(self) -> List[np.dtype]:
-        """ Convenience shortcut to get the datatypes as pandas types."""
-        return [x.type.to_pandas() for x in self.columns]
+        """ Convenience shortcut to get the datatypes as pandas types. Unsupported by TensorSpec."""
+        if self.is_tensor_spec():
+            raise MlflowException("TensorSpec only supports numpy types, use numpy_types() instead")
+        return [x.type.to_pandas() for x in self.inputs]
 
     def as_spark_schema(self):
         """Convert to Spark schema. If this schema is a single unnamed column, it is converted
         directly the corresponding spark data type, otherwise it's returned as a struct (missing
         column names are filled with an integer sequence).
+        Unsupported by TensorSpec.
         """
-        if len(self.columns) == 1 and self.columns[0].name is None:
-            return self.columns[0].type.to_spark()
+        if self.is_tensor_spec():
+            raise MlflowException("TensorSpec cannot be converted to spark dataframe")
+        if len(self.inputs) == 1 and self.inputs[0].name is None:
+            return self.inputs[0].type.to_spark()
         from pyspark.sql.types import StructType, StructField
 
         return StructType(
             [
                 StructField(name=col.name or str(i), dataType=col.type.to_spark())
-                for i, col in enumerate(self.columns)
+                for i, col in enumerate(self.inputs)
             ]
         )
 
     def to_json(self) -> str:
         """Serialize into json string."""
-        return json.dumps([x.to_dict() for x in self.columns])
+        return json.dumps([x.to_dict() for x in self.inputs])
 
     def to_dict(self) -> List[Dict[str, Any]]:
         """Serialize into a jsonable dictionary."""
-        return [x.to_dict() for x in self.columns]
+        return [x.to_dict() for x in self.inputs]
 
     @classmethod
     def from_json(cls, json_str: str):
         """ Deserialize from a json string."""
-        return cls([ColSpec(**x) for x in json.loads(json_str)])
```

Comment:
The json for tensors is supposed to look like this:
```
inputs: [{
  "name": "image",
  “type”: “tensor”,
  “tensor-spec“: {
    “dtype”: “float32”,
    “shape”: [28, 28, 3]
  }
}],
outputs: [{
  # probabilities:  
  “type”: “tensor”,
  “tensor-spec”: {
    “dtype”: “float32”,
    “shape”: [3] 
   }
}]
```

so we need to change this part a bit.